### DISCUSSION TITLE
Automatic python library download using pip does not work

### DISCUSSION THREAD

**Discussion by hrntsm** [Tunny-gh/Tunny_in_Open_Sea#154] [General]
From Food4Rhino reviews

https://www.food4rhino.com/en/comment/39042#comment-39042

> Hi, I am having trouble getting the Optimizer running.
> When I double-click the Tunny component I get the expected Python Installer pop up (its says the process runs when Tunny is launched the first time but it runs every single time I double click the component). After it runs and says "Finish!!" I close that out and run optimize but, I get another pop up; Tunny runtime error, that states " No module named 'Optuna' ". It also states that the solver will not work after this error unless Rhino is restarted, which I've done many times now. I've also tried to uninstall and re-install to no resolve. When I close the error message I can only click on the tabs of the solver interface for a little while before the entire interface is unresponsive and will not even close out.. then I have to force close rhino-gh. I have even pip installed Optuna separately and checked that it is installed and Tunny still gives the same error.
> 
> I am unsure why I am getting the error not sure if it is because I already had python and some of the libraries utilized by Tunny installed on my computer - but those are located outside of the gh Components folder so I don't know what the problem is. Thank you for your assistance in finding a solution. Thanks!

**Comment by hrntsm** [comment]
First, Thanks for using Tunny. Questions answered.

Q：
> its says the process runs when Tunny is launched the first time but it runs every single time I double click the component

A：

A bug in PythonInstaller can cause it to start every time, even if it is installed properly. This will be fixed in the next version.

Q：
>Tunny runtime error, that states " No module named 'Optuna' "......

A：

Could you check if the folder in which Tunny.gha is installed has the following structure?  
The "python-3.10.0-embed-amd64 folder" and "python-3.10.0-embed-amd64.zip" are downloaded by Tunny's PythonInstaller, and "Settings.json" is create when Tunny window wake up, so it is not initially included.

<img width="654" alt="Screenshot 2023-03-20 at 16 23 41" src="https://user-images.githubusercontent.com/23289252/226272805-1223d6c6-92ed-4f64-925d-e7cf3a557719.png">

When PythonInstaller is fully working, you will find several libraries in the site-packages folder, as shown below. This includes the "Optuna" folder.

<img width="633" alt="Screenshot 2023-03-20 at 16 24 14" src="https://user-images.githubusercontent.com/23289252/226272919-bee604f0-9fa1-4c10-88ce-26a094f853c7.png">

Tunny downloads these files automatically, and this 'automatic' process may cause the download to fail if the PC is highly secure.

**Reply by ChiaOP** [reply]
Hey, thank you very much for your response, and for creating this plug-in.

Regarding the second answer: I do have that file folder structure.
A.  I have the python embed folders. and "settings.json" is downloaded there (I guess this is because the python installer has already run numerous times).
![Screenshot 2023-03-20 211848](https://user-images.githubusercontent.com/90870444/226498267-db5fdac6-c5a8-4d9a-bb11-3c0c7f284924.png)

B. However, "site-packages" folder is empty
![image](https://user-images.githubusercontent.com/90870444/226498786-2fe57ccd-7058-4ead-ac43-620b491ab958.png)

I've checked to see if there are restrictions in security settings etc., and there are none that I can find.

Question:
If I receive all the folders for the various libraries from you, download them, and put them in the "site-packages" folder manually, do you think it will work?

**Reply by hrntsm** [reply]
Thanks for the confirmation.
It seems that the library download is not working.

It will work with a zip file of the data, but some features will not work.
For example, the dashboard function that displays optimization results will not work.

If you are familiar with the CLI, you can also try downloading the library manually by.

1. Get "get-pip.py" file from the following link
    - https://bootstrap.pypa.io/get-pip.py
1. Place the retrieved file in Lib as follows. The "site-packages" will be created automatically after this, so there is no need to create them.
    <img width="165" alt="image" src="https://user-images.githubusercontent.com/23289252/226514990-b2baeccc-0cd2-4fe8-bf89-9f2ab0fe4568.png">
1. Access the Tunny's folder "python-3.10.0-embed-amd64" containing "python.exe" with cmd or powershell and execute the following commands. This will allow you to download the pip to be used for installing the library.
    <img width="423" alt="image" src="https://user-images.githubusercontent.com/23289252/226516010-5fd5d56d-24e0-4996-a74c-1c84acde7bb6.png">
    ```bash
    ./python.exe .\Lib\get-pip.py
    ```
1. Next, install the libraries that should be in "site-packages" using the pip installed. If this works without problems, the necessary libraries are downloaded in "site-packages".
    ```bash
    ./python.exe -m pip install -r ..\Lib\requirements.txt
    ```

**Reply by ChiaOP** [reply]
Great, that worked. the packages downloaded.

![Tunny_Lib_Py_Libraries](https://user-images.githubusercontent.com/90870444/226536199-5e4079ac-57d5-4315-8e0a-616952a4b920.png)

I restarted the computer and programs and tried to run the Tunny optimization component in my gh script and now I am getting a sqlite error

![Tunny_Sqllite3_error](https://user-images.githubusercontent.com/90870444/226536215-6d67e925-867b-40c0-a352-6c5600d69073.png)

I visited the webpage referenced in the error message I think I understand what is being suggested. However I think it requires access to edit and compile any code to call the db file or api, if it is something I can do from the CLI, what would be the necessary commands and links referenced please?

This is the webpage that opens when I put in the url prompt. [https://docs.sqlalchemy.org/en/14/errors.html#error-e3q8](url)

**Reply by ChiaOP** [reply]
I do see the sqlite py file and .dll file in the "python-3.10.0-embed-amd64" folder
![sqlFilesInPyFolder](https://user-images.githubusercontent.com/90870444/226538251-8c407832-e328-4063-a7f3-ee095aca6e6b.png)

**Reply by hrntsm** [reply]
First of all, I am glad that you were able to do this with the download.
It seems that you are not doing well in saving the optimised file, so specifying the path from the following part may work.
Also, there were some cases where an error occurred when using sqlite, so v0.7.0, released today, supports the JournalStorage format, which does not use sqlite.
You will have to install it again, including python, but you can try this as well.

![image](https://user-images.githubusercontent.com/23289252/226562239-d9f1949d-2902-499c-928f-11da6cc9df77.png)

Tunny v0.7.0
https://github.com/hrntsm/Tunny/releases/tag/v0.7.0

It's still not well developed and some parts don't work properly, but thanks again for using it!

**Reply by ChiaOP** [reply]
Thanks very much for your help and all you've done with your development. 
I did all of the above and the optimization is running now.

All the best

**Reply by ChiaOP** [reply]
I know you said it wouldn't work with me downloading the py files manually. But Is there anything I can do to get the dashboard working. 

when I click on open dashboard this is what I receive

![Tunny_Dash](https://user-images.githubusercontent.com/90870444/226725209-0c33880f-a9a1-422c-aa31-2de4e3a8b0cd.png)
![Tunny_Dash_loading](https://user-images.githubusercontent.com/90870444/226725451-0c6e9c53-2dbf-413d-bcaf-8d172ea6faf1.png)




I have results in log but I can't get the data in the gh interface or  visualize optimal solutions

![Tunny_Run](https://user-images.githubusercontent.com/90870444/226726951-81c49601-22da-4cdf-9e86-bb448e80e725.png)





How do I know which is optimal?  I have no clue how to know which set in the log are the most optimal as I get this sql error when trying to run show the visualizations 

![CAN'T SAVE VIZ](https://user-images.githubusercontent.com/90870444/226728341-8b6de538-47b5-43d6-b91e-8c056eb9d73b.png)

**Reply by hrntsm** [reply]
> I know you said it wouldn't work with me downloading the py files manually. But Is there anything I can do to get the dashboard working.
when I click on open dashboard this is what I receive

The dashboard appears to be working correctly according to the images I have received.
Listening on .... as shown in the following image and then access http://127.0.0.1:8080/ with a browser, the dashboard should appear.

<img width="1253" alt="image" src="https://user-images.githubusercontent.com/23289252/226777972-3acc03ed-8d0b-4d11-8e95-a32307811dc8.png">

> How do I know which is optimal? I have no clue how to know which set in the log are the most optimal as I get this sql error when trying to run show the visualizations

The reflection of the results to gh does not seem to be working properly due to the support for non-sqlite since v0.7.0. This is a bug and will be fixed.

**Reply by hrntsm** [reply]
I created issue #161

**Reply by hrntsm** [reply]
I have made a fixed version, can you try it here?
You can download it from Artifacts at the bottom of the page.

https://github.com/hrntsm/Tunny/actions/runs/4485722493
